### PR TITLE
Remove value length limit from HTML parser & interpolated alignments

### PR DESCRIPTION
**Pull request by jelmervdl**
The HTML parser used internal stack-allocated buffers for text, tags, and attributes. This change replaces them with std::string_view.

This is a fix for #273.

Related issue in warc2text: https://github.com/bitextor/warc2text/issues/32, but less impact because that doesn't use these buffers for text between tags as it does entity decoding at another place.

This change also removes the fallback tag alignment code. Now if there is no alignment information available, no HTML tags will be inserted into the output. This would have already been enforced by #271 by always requesting alignment info from the model if ResponseOptions.HTML was true, so there should be no impact outside of the HTML class and tests. (And if there is the integration tests will call me on it.)

Edit: furthermore, this pulled in some tiny improvements:
- Jerin requested earlier that the lifetime of `new Tag` was a bit better explained
- I noticed that end tokens of sentences were not always aligned with each other if you just go by score. That's now enforced.
- Bit more consistency in using `numSentences()` and `numWords()` instead of suddenly jumping over to `alignments[..].size()`.

### CODE REVIEW

**Comment by kpu** [comment]
`xh_scanner` was designed with a rolling buffer in mind.  But here we have it in memory.  Can we use `string_view` instead?

**Comment by jelmervdl** [comment]
It now relies on buffers because of entities decoding. But if I emit TT_TEXT for entities separately, and also split emitting attr name and attr value instead of just TT_ATTR, I think I could replace all of the internal buffers with just string_view pointing to the original source or a hard-coded decoded entity.

File: src/translator/xh_scanner.h [hunk 1]
```diff
@@ -5,17 +5,41 @@
...
+  explicit instream(const char *src) : p(src), begin(src), end(src + strlen(src)) {}
+  instream(const char *begin, const char *end) : p(begin), begin(begin), end(end) {}
+  char consume() { return p < end ? *p++ : 0; }
+  char peek() const { return p < end ? *p : 0; }
+  const char *pos() const { return p; }
+};
+
+// Very simple string_view implementation that has mutable size so we can easily
+// grow or shrink it while we scan through the input.
+struct string_view {
```
[the remark this answers is not in the document]
Comment (jelmervdl) [reply src/translator/xh_scanner.h]: I needed something that has a mutable size, but I agree having a third implementation of the same concept is stupid.

I've now renamed the struct to `string_ref` to be more different, removed the utility functions, and removed it from the public `xh_scanner` interface. That now just exposes `std::string_view` (and converts my mutable references to `std::string_view` when accessed.)

Only annoyance I've encountered with `std::string_view` is doing lookups in a `std::unordered_map<std::string>`. I convert them on the fly, which seems wasteful, but the `""sv` string literals did not work for me, and hopefully short string optimisation will remove the need for tiny allocations.

**Comment by andrenatal** [comment]
Is it possible to expand test coverage and be sure this functionality is stable and covering edge cases?

**Comment by jerinphilip** [comment]
> This change also removes the fallback tag alignment code. Now if there is no alignment information available, no HTML tags will be inserted into the output.

This should be a crash. `ABORT_IF(alignmentsMissing, ....)`. If somehow during the course of development or something we end up at a scenario where HTML is demanded and no alignments, it's a configuration error. I read this message, I should immediately know what to do.

We will surely make configuration easy for Mozilla with perhaps only one switch, but there should not be a correction code that provides a false pretence of things being okay.

**Comment by jelmervdl** [comment]
That's what I normally would do (fail early, fail often) but I was hesitant on introducing new ways to fail hard in bergamot-translator. But if you say go down that route, I will!

I'll add the ABORT_IF and rework/remove test cases that need changes due to this.

**Review by jerinphilip — APPROVED** [review]
https://github.com/jerinphilip/tagtransfer/pull/1 has run the bits on a bit of Wikipedia pages to detect a few more corner cases. Behind, we'll try more link-surfing.

Looks good on some basic checks.
